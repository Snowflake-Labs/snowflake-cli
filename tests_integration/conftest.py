from __future__ import annotations

import functools
import pytest
import toml

from pathlib import Path
from snowcli.cli import app
from tempfile import NamedTemporaryFile
from typer import Typer
from typer.testing import CliRunner

TEST_DIR = Path(__file__).parent


@pytest.fixture(scope="session")
def test_snowcli_config():
    test_config = TEST_DIR / "config/test.toml"
    config = toml.load(test_config)
    config["snowsql_config_path"] = str(TEST_DIR / "config/connection_configs.toml")
    with NamedTemporaryFile(suffix=".toml", mode="w+") as fh:
        toml.dump(config, fh)
        fh.flush()
        yield fh.name


class SnowCLIRunner(CliRunner):
    def __init__(self, app: Typer, test_snowcli_config: str):
        super().__init__()
        self.app = app
        self.test_snowcli_config = test_snowcli_config

    @functools.wraps(CliRunner.invoke)
    def invoke(self, *a, **kw):
        return super().invoke(self.app, *a, **kw)

    def invoke_with_config(self, *args, **kwargs):
        return self.invoke(
            ["--config-file", self.test_snowcli_config, *args[0]], **kwargs
        )


@pytest.fixture
def runner(test_snowcli_config):
    return SnowCLIRunner(app, test_snowcli_config)
