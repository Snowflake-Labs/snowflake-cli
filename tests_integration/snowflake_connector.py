from __future__ import annotations

import os
import pytest
import uuid

from snowflake import connector
from snowcli.exception import EnvironmentVariableNotFoundError

_ENV_PARAMETER_PREFIX = "SNOWFLAKE_CONNECTIONS_INTEGRATION"


@pytest.fixture(scope="session")
def create_database():
    database_name = f"db_{uuid.uuid4().hex}"

    config = {
        "application": "INTEGRATION_TEST",
        "account": _get_from_env("ACCOUNT", True),
        "user": _get_from_env("USER", True),
        "password": _get_from_env("PASSWORD", True),
        "schema": _get_from_env("SCHEMA", False, "PUBLIC"),
        "host": _get_from_env("HOST", False),
        "role": _get_from_env("ROLE", False),
    }
    config = {k: v for k, v in config.items() if v is not None}
    connection = connector.connect(**config)

    connection.execute_string(f"create database {database_name}")
    os.environ[f"{_ENV_PARAMETER_PREFIX}_DATABASE"] = database_name

    yield database_name

    connection.execute_string(f"drop database {database_name}")
    del os.environ[f"{_ENV_PARAMETER_PREFIX}_DATABASE"]


@pytest.fixture(scope="session")
def snowflake_session(create_database):
    config = {
        "application": "INTEGRATION_TEST",
        "account": _get_from_env("ACCOUNT", True),
        "user": _get_from_env("USER", True),
        "password": _get_from_env("PASSWORD", True),
        "host": _get_from_env("HOST", False),
        "database": _get_from_env("DATABASE", False),
        "role": _get_from_env("ROLE", False),
        "schema": "PUBLIC",
        "warehouse": _get_from_env("WAREHOUSE", False),
    }
    config = {k: v for k, v in config.items() if v is not None}
    connection = connector.connect(**config)
    yield connection
    connection.close()


def _get_from_env(parameter_name: str, required: bool, default=None) -> str | None:
    env_value = os.environ.get(f"{_ENV_PARAMETER_PREFIX}_{parameter_name}")
    if required and not env_value:
        if default is None:
            raise EnvironmentVariableNotFoundError(
                f"{_ENV_PARAMETER_PREFIX}_{parameter_name}"
            ) from None
        return default
    return env_value
