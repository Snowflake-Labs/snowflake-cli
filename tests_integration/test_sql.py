import pytest

from unittest import mock
from tests_integration.test_utils import extract_data, extract_list_of_data
from tests_integration.snowflake_connector import snowflake_session


@pytest.mark.integration
@mock.patch("snowcli.cli.sql.print_db_cursor")
def test_query_parameter(mock_print, runner, snowflake_session):
    runner.invoke_with_config(["sql", "-q", "select pi()"])

    assert extract_data(mock_print) == [{"PI()": 3.141592654}]


@pytest.mark.integration
@mock.patch("snowcli.cli.sql.print_db_cursor")
def test_multi_queries_from_file(mock_print, runner, snowflake_session, test_root_path):
    runner.invoke_with_config(
        ["sql", "-f", f"{test_root_path}/test_files/sql_multi_queries.sql"]
    )

    assert extract_list_of_data(mock_print) == [
        [{"LN(1)": 0}],
        [{"LN(10)": 2.302585093}],
        [{"LN(100)": 4.605170186}],
    ]
