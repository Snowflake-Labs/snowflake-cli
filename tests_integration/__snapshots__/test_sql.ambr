# serializer version: 1
# name: test_if_comments_are_not_shown
  '''
  SELECT 1;
  +---+
  | 1 |
  |---|
  | 1 |
  +---+
  
  '''
# ---
# name: test_if_comments_flag_works
  '''
  -- this is a comment;
  SELECT 1;
  +---+
  | 1 |
  |---|
  | 1 |
  +---+
  
  '''
# ---
